// @node-red/nodes/core/meo/20-meo-command.js
module.exports = function(RED) {
    "use strict";

    var caps = require("./lib/meo-caps");

    // The gateway waits 10s for the device's reply (MeoControlHandler
    // REPLY_TIMEOUT_MS); give its own response room to arrive after that.
    var COMMAND_TIMEOUT = 15000;

    function MeoCommandNode(n) {
        RED.nodes.createNode(this, n);
        var node = this;
        node.gateway = RED.nodes.getNode(n.gateway);
        node.device = n.device;
        node.cap = Number(n.cap);
        node.value = n.value;
        node.valueType = n.valueType || "msg";

        node.on("input", function(msg, send, done) {
            if (!node.gateway) {
                fail(node, done, "meo.errors.noGateway");
                return;
            }
            if (!node.device) {
                fail(node, done, "meo.errors.noDevice");
                return;
            }
            if (!(node.cap > 0)) {
                fail(node, done, "meo.errors.noCap");
                return;
            }

            // Only writes carry a value; the gateway ignores it for reads and
            // generic commands (docs/mqtt_messaging.md "Command").
            if (caps.kindOf(node.cap) !== "write") {
                sendCommand(node, msg, send, done, 0);
                return;
            }
            RED.util.evaluateNodeProperty(node.value, node.valueType, node, msg, function(err, value) {
                if (err) {
                    fail(node, done, "meo.errors.badValue", err);
                    return;
                }
                var scalar = toInt32(value);
                if (scalar === null) {
                    fail(node, done, "meo.errors.badValue",
                        new Error(RED._("meo.errors.notANumber", { value: String(value) })));
                    return;
                }
                sendCommand(node, msg, send, done, scalar);
            });
        });
    }

    function sendCommand(node, msg, send, done, value) {
        var url = node.gateway.apiUrl() + "/api/v1/devices/" +
            encodeURIComponent(node.device) + "/command";

        node.status({ fill: "blue", shape: "dot", text: "node-red:meo.status.sending" });

        fetch(url, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify({ cap: node.cap, value: value }),
            signal: AbortSignal.timeout(COMMAND_TIMEOUT)
        }).then(function(response) {
            // A failed command still answers with a JSON body worth reading.
            return response.json().catch(function() { return {}; }).then(function(body) {
                return { ok: response.ok, status: response.status, body: body };
            });
        }).then(function(result) {
            if (!result.ok) {
                throw new Error(explain(result.status, result.body));
            }
            msg.payload = result.body.value;
            msg.deviceId = result.body.deviceId;
            msg.cap = result.body.cap;
            node.status({ fill: "green", shape: "dot", text: String(result.body.value) });
            send(msg);
            done();
        }).catch(function(err) {
            // AbortSignal.timeout rejects with a TimeoutError, not our message.
            var reason = err.name === "TimeoutError"
                ? RED._("meo.errors.gatewayTimeout")
                : err.message;
            node.status({ fill: "red", shape: "ring", text: reason });
            done(new Error(reason));
        });
    }

    // Turn the gateway's status codes (ControlController.statusFor) into
    // something a classroom can act on. Its own message is already plain
    // English, so only the two most confusing cases get reworded.
    function explain(status, body) {
        if (status === 404) { return RED._("meo.errors.deviceNotFound"); }
        if (status === 504) { return RED._("meo.errors.noReply"); }
        return body.error || RED._("meo.errors.gatewayStatus", { status: status });
    }

    // The API takes a single int32. Booleans are the common inject payload for
    // an on/off actuator, so treat them as 1/0 rather than rejecting them.
    function toInt32(value) {
        if (typeof value === "boolean") { return value ? 1 : 0; }
        if (value === null || value === undefined || value === "") { return null; }
        var number = Number(value);
        if (!isFinite(number)) { return null; }
        return Math.round(number);
    }

    // Status text is resolved by the editor, so it carries the namespace;
    // the error message is resolved here and does not.
    function fail(node, done, key, err) {
        node.status({ fill: "red", shape: "ring", text: "node-red:" + key });
        done(new Error(err ? err.message : RED._(key)));
    }

    RED.nodes.registerType("meo-command", MeoCommandNode);
};
